umapi_client/functional.py

```diff
@@ -308,6 +309,61 @@ def delete_account(self):
         self.append(removeFromDomain={})
         return None
 
+    def maybe_split_groups(self, max_groups):
+        """
+        Check if group lists in add/remove directives should be split and split them if needed
+        :param max_groups: Max group list size
+        :return: True if at least one command was split, False if none were split
+        """
+        new_commands = []
+        # return True if we split at least once
+        maybe_split = False
+        for command in self.commands:
+            do_split = False
+            for verb, verb_commands in six.iteritems(command):
```

Comment:
Do we always know that a command is a dictionary of one key?  We can probably assume so, since the UserAction interface appears to only add one key per command.  But it is perfectly valid to do something like this:

```
[
  {
    "user": "testuser1@ccestestdomain.com",
    "requestID": "action_1",
    "do": [
      {
        "createEnterpriseID": {
          "email": "testuser1@ccestestdomain.com",
          "option": "ignoreIfAlreadyExists",
          "country": "US"
        }
      },
      {
        "add": {
          "usergroup": [
            "too many groups 11",
            "too many groups 4",
            "too many groups 5",
            "too many groups 14",
            "too many groups 12",
            "too many groups 9",
            "too many groups 10",
            "too many groups 2",
            "too many groups 3",
            "too many groups 1"
          ],
          "product": [
            "Acrobat Pro DC"
          ]
        },
        "remove": {
          "usergroup": [
            "too many groups 15"
          ]
        }
      }
    ]
  }
]
```

You'll notice that the second command calls add and remove.  I figured my split code should follow what the API will allow, even if the UserAction may not structure a command that way.  I can have it support a single key if that would be better.

Reply:
OK, so the terminology of the API spec is very confusing.  The top-level dictionary is the "command", and the operation dictionaries in the command's "do" list are called "steps".  (Unfortunately, I got the language wrong when I wrote umapi-client, and I called the top-level dictionary the "action", and the each dictionary in the "do" list I call a "command".  So the function I call "split_action", which limits each "action" to 10 "commands", should really have been called "split_command" as it limits each action to 10 "steps".  Sigh.

No matter which terminology you use, it's clear in the API spec that each "step" in the "do list" is only allowed to have one "step action" (what we are calling a verb), and that the "do list" is only allowed to have 10 "step actions".  The form you have specified above is, in fact, not allowed by the API spec, because it would (indirectly) allow more than 10 step actions per command (by allowing one step to have multiple step actions).  In addition, because the structure is a dict, and the order of keys in a dict is not guaranteed, you would have no way of knowing what order these operations are executed in, which would be a terrible API spec.

Unfortunately, I was just looking at the command validation code on the server, and though it has several comments stating that there should be just one step action per step, I'm not convinced it enforces that.  In fact, I'm pretty sure it doesn't.

Sigh.  What to do.  We could enforce one step action per step in the umapi-client, which would be slightly backward-incompatible, but is the only thing that makes any sense, because processing all the step actions and then breaking out each step action with long arguments into a separate step (which is what we are doing) would completely break the step action ordering if there were more than one.

So let's not assume there is just one step action per step, let's enforce that, and then we don't need to do any deep copies.  And I'll follow up with the server team to find out why they aren't enforcing one step action per step.

Reply:
And, by the way, there is an incorrect example (with multiple step actions per step) also in the doc of UserGroup actions.  I'll get that fixed as well.

Reply:
That makes sense.  I'll use next() to get the first (and presumably) only key from the step.

Reply:
Sounds great.  Please also report an error if there is more than one.